pkg/schedulerplugin/scheduler_plugin.go

```diff
@@ -0,0 +1,209 @@
+//
+// Finally, in PostBind(), the allocation is removed as we now know that the pod has been allocated successfully.
+// If a pending or in-progress allocation is detected for a pod in PreFilter(), we remove the allocation and force the
+// pod to be rescheduled, as this means the prior allocation could not be completed successfully by the default
+// scheduler for some reason.
+type YuniKornSchedulerPlugin struct {
+	sync.RWMutex
+	context *cache.Context
+}
+
+// ensure all required interfaces are implemented
+var _ framework.PreFilterPlugin = &YuniKornSchedulerPlugin{}
+var _ framework.FilterPlugin = &YuniKornSchedulerPlugin{}
+var _ framework.PostBindPlugin = &YuniKornSchedulerPlugin{}
+
+// Name returns the name of the plugin
+func (sp *YuniKornSchedulerPlugin) Name() string {
+	return SchedulerPluginName
+}
+
+// PreFilter is used to release pods to scheduler
+func (sp *YuniKornSchedulerPlugin) PreFilter(_ context.Context, _ *framework.CycleState, pod *v1.Pod) *framework.Status {
+	log.Logger().Debug("PreFilter check",
+		zap.String("pod", fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)))
+
+	// we don't process pods without appID defined
+	appID, err := utils.GetApplicationIDFromPod(pod)
+	if err != nil {
+		log.Logger().Info(fmt.Sprintf("Skipping pod %s/%s in the prefilter plugin because no applicationID is defined",
+			pod.Namespace, pod.Name))
+		return framework.NewStatus(framework.Success, "Deferring to default scheduler")
+	}
+
+	if app := sp.context.GetApplication(appID); app != nil {
+		if task, err := app.GetTask(string(pod.UID)); err == nil {
+			_, ok := sp.context.GetInProgressPodAllocation(string(pod.UID))
+			if ok {
+				// pod must have failed scheduling, reject it and return unschedulable
+				log.Logger().Info("Task failed scheduling, marking as rejected",
+					zap.String("pod", fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)),
+					zap.String("taskID", task.GetTaskID()))
+				sp.context.RemovePodAllocation(string(pod.UID))
+				dispatcher.Dispatch(cache.NewRejectTaskEvent(app.GetApplicationID(), task.GetTaskID(),
+					fmt.Sprintf("task %s rejected by scheduler", task.GetTaskID())))
+				return framework.NewStatus(framework.Unschedulable, "Pod is not ready for scheduling")
+			}
+
+			nodeID, ok := sp.context.GetPendingPodAllocation(string(pod.UID))
+			if task.GetTaskState() == events.States().Task.Bound && ok {
+				log.Logger().Info("Releasing pod for scheduling (prefilter phase)",
+					zap.String("pod", fmt.Sprintf("%s/%s", pod.Namespace, pod.Name)),
+					zap.String("taskID", task.GetTaskID()),
+					zap.String("assignedNode", nodeID))
+
+				return framework.NewStatus(framework.Success, "")
+			}
+		}
+	}
+
+	return framework.NewStatus(framework.Unschedulable, "Pod is not ready for scheduling")
+}
+
+// PreFilterExtensions is unused
+func (sp *YuniKornSchedulerPlugin) PreFilterExtensions() framework.PreFilterExtensions {
+	return nil
+}
+
+// Filter is used to release specific pod/node combinations to scheduler
```

Comment:
The thing that's probably worth mentioning here is what happens when `expectedNode != node`, pointing out the minimal performance impact and the immediate invocation of `Filter()` again. You can probably also add that currently this is the only way to force the default scheduler to match YK's decision.

Reply:
Added comments.